### ISSUE TITLE
`ref.exists()` returns `true` when manually invalidate a provider.

### ISSUE THREAD

**Issue by hasanmhallak** [rrousselGit/riverpod#3558]
**Describe the bug**
It appears that `ref.exist(someProvider)` still returns `true` when manually invalidate a provider, and when try to use the invalidated provider `onResume` gets called. While when we set `keepAlive` to `true`  or use `AutoDisposeNotifier` everything works as expected.

This behavior is found when using / not using generated code.  

Original issue #2729 

**To Reproduce**

<details> <Summary>Code Generation Sample</Summary>

```dart

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';
part 'main.g.dart';

@Riverpod(keepAlive: true)
int fetchItem(FetchItemRef ref) {
  // this will be called on invalidate.
  ref.onDispose(() => print('dispose'));

  ref.onCancel(() => print('onCancel'));

  // but this will be called when we use
  // the provider after invalidating
  ref.onResume(() => print('onResume'));
  return 5;
}

void main() {
  runApp(
    const ProviderScope(child: MaterialApp(home: MyHomePage())),
  );
}

class MyHomePage extends ConsumerWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      floatingActionButton: FloatingActionButton(onPressed: () {
        print('does exist: ${ref.exists(fetchItemProvider)}');
      }),
      body: Center(
        child: TextButton(
          onPressed: () {
            print('Navigate To Next Page');
            Navigator.of(context).push(MaterialPageRoute(builder: (_) => MyHomePage2()));
          },
          child: Text('Navigate To Next Page'),
        ),
      ),
    );
  }
}

class MyHomePage2 extends ConsumerStatefulWidget {
  const MyHomePage2({super.key});

  @override
  ConsumerState<MyHomePage2> createState() => _MyHomePage2State();
}

class _MyHomePage2State extends ConsumerState<MyHomePage2> {
  ProviderContainer? _container;

  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    _container!.invalidate(fetchItemProvider);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    _container = ProviderScope.containerOf(context);
    return Scaffold(
      appBar: AppBar(),
      floatingActionButton: FloatingActionButton(onPressed: () {
        print('does exist: ${ref.exists(fetchItemProvider)}');
      }),
      body: Center(
        child: Text(ref.watch(fetchItemProvider).toString()),
      ),
    );
  }
}
```
</details>

<details> <Summary>No Code Generation Sample with explanation</Summary>

**To Reproduce** 

1. run the code
2. Click Exist Fab
3. Navigate to next page
4. Click the + Fab to increment state
5. Click Exist
6. Navigate Back
7. Click Exist

**Logs**
```console
flutter: does exist: false
flutter: internal State: 5
flutter: Navigate To Next Page
flutter: does exist: true
flutter: internal State: 7
flutter: dispose
flutter: does exist: true // the instance still exists and the internal state is preserved.
flutter: internal State: 7
```

```dart

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

class FetchItemsNotifier extends Notifier<int> {
  int internalState = 5;
  @override
  build() {
    // this will be called on invalidate.
    ref.onDispose(() => print('dispose'));

    ref.onCancel(() => print('onCancel'));

    // but this will be called when we use
    // the provider after invalidating
    ref.onResume(() => print('onResume'));
    return 5;
  }

  void add() {
    state = state + 1;
    internalState = state;
  }
}

final fetchItemProvider = NotifierProvider<FetchItemsNotifier, int>(FetchItemsNotifier.new);

void main() {
  runApp(
    const ProviderScope(child: MaterialApp(home: MyHomePage())),
  );
}

class MyHomePage extends ConsumerWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      floatingActionButton: FloatingActionButton(
        child: Text('Exist'),
        onPressed: () {
          print('does exist: ${ref.exists(fetchItemProvider)}');
          print('internal State: ${ref.read(fetchItemProvider.notifier).internalState}');
        },
      ),
      body: Center(
        child: TextButton(
          onPressed: () {
            print('Navigate To Next Page');
            Navigator.of(context).push(MaterialPageRoute(builder: (_) => MyHomePage2()));
          },
          child: Text('Navigate To Next Page'),
        ),
      ),
    );
  }
}

class MyHomePage2 extends ConsumerStatefulWidget {
  const MyHomePage2({super.key});

  @override
  ConsumerState<MyHomePage2> createState() => _MyHomePage2State();
}

class _MyHomePage2State extends ConsumerState<MyHomePage2> {
  ProviderContainer? _container;

  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    _container!.invalidate(fetchItemProvider);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    _container = ProviderScope.containerOf(context);
    return Scaffold(
      appBar: AppBar(),
      floatingActionButton: Row(
        mainAxisAlignment: MainAxisAlignment.end,
        children: [
          FloatingActionButton(
            heroTag: 'Exist',
            child: Text('Exist'),
            onPressed: () {
              print('does exist: ${ref.exists(fetchItemProvider)}');
              print('internal State: ${ref.read(fetchItemProvider.notifier).internalState}');
            },
          ),
          FloatingActionButton(
            child: Text('+'),
            onPressed: () {
              ref.read(fetchItemProvider.notifier).add();
            },
          ),
        ],
      ),
      body: Center(
        child: Text(ref.watch(fetchItemProvider).toString()),
      ),
    );
  }
}

```

When we switch to using `AutoDisposeNotifier` the internal state is resets (which means the old provider instance are removed from the container) and `ref.exists()` return `false`. which will produce the following logs:  

**Logs When Using `AutoDisposeNotifier` **
```console
flutter: does exist: false
flutter: internal State: 5
flutter: dispose // got disposed after accessing the internal state of the provider when no active listener registered.
flutter: Navigate To Next Page
flutter: does exist: true
flutter: internal State: 8
flutter: dispose 
flutter: does exist: false // auto dispose works as expected.
flutter: internal State: 5 // a new AutoDisposeNotifier instance created
flutter: dispose // got disposed after accessing the internal state of the provider when no active listener registered.
```
</details>

**Comment by lucavenir** [comment]
Hello 👋🏼

I am *not* able to replicate this via the following test

<details><summary>self-enclosing test repro</summary>
note: tested with riverpod 3.0.0-dev15

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_persist_problem/observer.dart';

class MyNotifier extends Notifier<int> {
  @protected
  late int internal;
  @override
  int build() {
    internal = 0;
    return internal;
  }

  void increment() {
    state++;
    internal = state;
  }
}

final myNotifierProvider = NotifierProvider.autoDispose<MyNotifier, int>(
  MyNotifier.new,
  name: 'myNotifierProvider',
);

void main() {
  test('ref.exists should be false on invalidated providers', () async {
    final container = ProviderContainer.test(observers: [const MyObserver()]);

    expect(container.exists(myNotifierProvider), isFalse);
    expect(container.read(myNotifierProvider), equals(0));
    final sub = container.listen(myNotifierProvider, (previous, next) {});
    expect(container.exists(myNotifierProvider), isTrue);
    container.read(myNotifierProvider.notifier).increment();
    container.read(myNotifierProvider.notifier).increment();
    expect(container.read(myNotifierProvider), equals(2));
    sub.close();
    await container.pump();
    expect(container.exists(myNotifierProvider), isFalse);
    expect(container.read(myNotifierProvider), equals(0));
  });
}
```
</details>

**Comment by hasanmhallak** [comment]
Hey @lucavenir, you did not create  an AutoDisposeNotifier.

You should extend `AutoDisposeNotifier` and use the `autoDispose` modifier when defining the provider.

**Comment by lucavenir** [comment]
Thanks!! My bad, got distracted. This way, the test passes. I'll think about another way to replicate this one with a test.

**Comment by lucavenir** [comment]
Hi again 👋🏼

I tried your code and I just encounter _expected behavior_.

- riverpod 2.6.1: your logs are expected - the notifier dispose and the internal state is preserved because `ref.invalidate` only affects `build`. The only unexpected thing for me is `onResume` not being triggered, but it's fixed in 3.x
- riverpod 3.0.0-dev15: my logs (below) are expected - in this version, internal state will be destroyed on invalidations (and indeed that happens). 
- riverpod 3.0 (future): the same behavior in 2.6.1 is to be expected

Then, I tweaked a little bit to give you a little more insights.

(1)
I *slightly* adjusted the `.invalidate` logic: you're not supposed to use `ProviderScope.containerOf` as a workaround of [this](https://pub.dev/packages/riverpod_lint#avoid_ref_inside_state_dispose).
Besides the above lint, using `ref` in `dispose` leads to a `StateError`.
Indeed, I've moved the `ref.invalidate` logic into `deactivate` instead.

(2)
You're building a `Notifier`, so it will stay alive even after this invalidation.
Its internal state *won't* reset, that's by design, because `ref.invalidate` only refreshes `build`, and won't destroy the object itself.
Notice how [this will be the case](https://github.com/rrousselGit/riverpod/pull/4135) for riverpod 3 as well.

(3)
If you test this with riverpod 3.0.0-dev15, you'd notice the internal state reset back to its state. That's because the above PR didn't land for that version. Indeed, check the logs:

<details><summary>my logs</summary>

```
does exist: false
onCancel
internal State: 5
Navigate To Next Page
onResume
does exist: true
internal State: 7
onDispose
does exist: true
onResume
onCancel
internal State: 5
```
</details>

Let me know if you're satisfied with this answer, or if I'm missing something.
I won't write tests until I'm sure there's something missing, here.

---

<details><summary>the code</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

class FetchItemsNotifier extends Notifier<int> {
  @protected
  var internalState = 5;
  @override
  build() {
    ref.onDispose(() => print('onDispose'));
    ref.onCancel(() => print('onCancel'));
    ref.onResume(() => print('onResume'));
    return 5;
  }

  void add() {
    state = state + 1;
    internalState = state;
  }
}

final fetchItemProvider = NotifierProvider<FetchItemsNotifier, int>(
  FetchItemsNotifier.new,
);

void main() {
  runApp(const ProviderScope(child: MaterialApp(home: MyHomePage())));
}

class MyHomePage extends ConsumerWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.select_all),
        onPressed: () {
          print('does exist: ${ref.exists(fetchItemProvider)}');
          print(
            'internal State: ${ref.read(fetchItemProvider.notifier).internalState}',
          );
        },
      ),
      body: Center(
        child: TextButton(
          onPressed: () {
            print('Navigate To Next Page');
            Navigator.of(
              context,
            ).push(MaterialPageRoute(builder: (_) => MyHomePage2()));
          },
          child: Text('Navigate To Next Page'),
        ),
      ),
    );
  }
}

class MyHomePage2 extends ConsumerStatefulWidget {
  const MyHomePage2({super.key});

  @override
  ConsumerState<MyHomePage2> createState() => _MyHomePage2State();
}

class _MyHomePage2State extends ConsumerState<MyHomePage2> {
  @override
  void deactivate() {
    ref.invalidate(fetchItemProvider);
    super.deactivate();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      floatingActionButton: Row(
        spacing: 16,
        mainAxisAlignment: MainAxisAlignment.end,
        children: [
          FloatingActionButton(
            heroTag: 'Exist',
            child: Icon(Icons.select_all),
            onPressed: () {
              print('does exist: ${ref.exists(fetchItemProvider)}');
              print(
                'internal State: ${ref.read(fetchItemProvider.notifier).internalState}',
              );
            },
          ),
          FloatingActionButton(
            child: Icon(Icons.add),
            onPressed: () => ref.read(fetchItemProvider.notifier).add(),
          ),
        ],
      ),
      body: Center(child: Text(ref.watch(fetchItemProvider).toString())),
    );
  }
}
```
</details>

**Comment by hasanmhallak** [comment]
Hey @lucavenir,

This issue is a follow-up to the original discussion in [#2729](https://github.com/rrousselGit/riverpod/issues/2729), where Remi mentioned in [this comment](https://github.com/rrousselGit/riverpod/issues/2729#issuecomment-1636713192) that `_ref.invalidate` will be updated to destroy the element of `keepAlive` providers if they no longer have any listeners.

Regarding the points you brought up, I'm not sure if those were simply observations on your part but just to clarify, using `ProviderScope.containerOf` is a standard way to access the container in a given context, and interacting with the container directly is the normal practice when working with Riverpod in pure Dart.

Doing the invalidation inside the `deactivate` method can be problematic. This method is triggered when a widget is removed from the tree—not necessarily when it's disposed. The widget might still be reinserted into the tree (for example, when reordering items list or when using a `GlobalKey`), which makes `deactivate` an unsafe place to perform clean-up logic like.

The core issue here is the inconsistency between manual and automatic provider disposal—especially considering that `invalidate` is currently treated as the primary method for disposing a provider, as discussed [here](https://github.com/rrousselGit/riverpod/issues/1295#issuecomment-1100625766).

Additional context and explanation can be found [here](https://github.com/rrousselGit/riverpod/issues/2729#issuecomment-1636710175).

---

**TL;DR**  
With `keepAlive: true`, using `invalidate()` does not fully dispose the provider or remove it from `_stateReaders`, leading to potential memory issues. This differs from the behavior of `autoDispose`, which handles cleanup correctly especially when using the `family` modifier to dynamically created provider for create-on-demand / infinite scroll ([#2728](https://github.com/rrousselGit/riverpod/issues/2728)).

**Comment by rrousselGit** [comment]
Going back to this: Calling `ref.invalidate` inside `State.dispose` is not legal at the moment.

There are a couple of things here:
- When inside `State.dispose`, subscriptions to ref.watch/ref.listen are not yet unsubscribed. So calling `ref.invalidate` there is no-op since the provider is still listened.
- Even if we close subscription _before_ calling `ref.invalidate`, calling  it inside `State.dispose` end-up throwing an assertion error (the typical "markNeedsBuild" error).

It's not really a bug, and rather just not supported at all.

It might be possible to support. Although this will need a couple of changes.

I'll raise a separate issue to have an official feature request, and close this bug report in favour of the new issue.
